client-core/src/service/hdkey_service.rs

```diff
@@ -0,0 +1,184 @@
+        }
+
+        Ok((public_key, private_key))
+    }
+
+    /// Retrieves private key corresponding to given public key
+    fn private_key(
+        &self,
+        public_key: &PublicKey,
+        passphrase: &SecUtf8,
+    ) -> Result<Option<PrivateKey>> {
+        let private_key_bytes =
+            self.storage
+                .get_secure(KEYSPACE, public_key.serialize(), passphrase)?;
+
+        private_key_bytes
+            .map(|mut private_key_bytes| {
+                let private_key = PrivateKey::deserialize_from(&private_key_bytes)?;
+                private_key_bytes.zeroize();
+                Ok(private_key)
+            })
+            .transpose()
+    }
+
+    /// Clears all storage
+    fn clear(&self) -> Result<()> {
+        self.storage.clear(KEYSPACE)
+    }
+}
+
+impl<T> HDKeyService<T>
+where
+    T: Storage,
+{
+    /// get random mnemonic
+    pub fn get_random_mnemonic(&self) -> String {
+        let mnemonic = Mnemonic::new(MnemonicType::Words24, Language::English);
+        mnemonic.clone().to_string()
+    }
+
+    /// generate seed from mnemonic
+    pub fn generate_seed(&self, mnemonic: &str, name: &str, passphrase: &SecUtf8) -> Result<()> {
+        debug!("hdwallet generate seed={}", mnemonic);
+        let mnemonic = Mnemonic::from_phrase(&mnemonic.to_string(), Language::English).unwrap();
+        let seed = Seed::new(&mnemonic, "");
+        self.storage
+            .set_secure(KEYSPACE, name, seed.as_bytes().into(), passphrase)?;
+        debug!("hdwallet write seed success");
+        Ok(())
+    }
+
+    /// auto-matically generate staking, transfer addresses
+    /// with just one api call
+    #[allow(dead_code)]
+    pub fn auto_restore(
+        &self,
+        mnemonic: &str,
+        name: &str,
+        passphrase: &SecUtf8,
+        count: i32,
+    ) -> Result<()> {
+        self.generate_seed(mnemonic, name, passphrase)
+            .expect("auto restore");
+        for index in 0..count {
+            let seed_bytes = self.storage.get_secure(KEYSPACE, name, passphrase)?;
+            for account in 0..2 {
+                let extended = ExtendedPrivKey::derive(
+                    &seed_bytes.clone().unwrap()[..],
+                    format!("m/44'/394'/{}'/0/{}", account, index).as_str(),
+                )
+                .unwrap();
+                let secret_key_bytes = extended.secret();
+
+                let private_key = PrivateKey::deserialize_from(&secret_key_bytes).unwrap();
+                let public_key = PublicKey::from(&private_key);
+
+                self.storage.set_secure(
+                    KEYSPACE,
+                    public_key.serialize(),
+                    private_key.serialize(),
+                    passphrase,
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Creates a new instance of key service
+    pub fn new(storage: T) -> Self {
+        HDKeyService { storage }
+    }
+
+    /// read value from db
```

Comment:
the comment isn't that useful

Reply:
it'll be better to explain what value it reads -- I assume `default` is what's returned if the value isn't found?

Reply:
added more explanation